update method for drawing slider bars

tgfrerer opened the pull request:
Instead of calling `ofPath` tessellation on each draw call for sliders, slider knows how to generate mesh for bar rectangles itself. The mesh is cached, and updated **iff** the slider value changes.

Rect mesh caching works similarly to the slider's internal caching of text meshes.

You might notice a slight performance improvement on GUIs with multiple sliders with this PR.

As an additional check/optimisation, slider fill rectangles with either width or height < 1 pixels will also not be drawn (they would not be visible anyway).

By *not* using `ofPath` (and by implication the tessellator) to draw the slider we avoid having to ask the tessellator to tessellate extremely small paths, in which case it had a tendency to throw its hands up in
despair or disgust (hard to tell from a stack trace) and cause crashes (see issue #6294).

Fixes issue #6294.

arturoc commented:
:) thanks! this looks actually really useful for other gui elements since pretty much everything is a rectangle in ofxGui and should be much faster to generate a mesh than tessellate a path. Can you move it outside of oxfSlider into utils or something like that? Perhaps just it's own file. No need to adapt other controls but we can start using it whenever we modify other classes

Also can you remove the allocation of the vectors on every change? given how small they are it's probably insignificant but better to avoid memory allocations. Just add the vertex info directly in the mesh.

roymacdonald commented:
Ohh thanks @tgfrerer ! :D
I will be able to look at this in depth next week and hopefully adapt it to the rest of ofxGui. I'll keep you posted. 
Cheers!

tgfrerer commented:
@arturoc I've updated the PR as to your suggestions. I've extracted the rectangle draw code into `ofxGuiUtils.h`, a separate header file which I have created in `ofxGUI`, where I felt it most likely belonged. 

I've placed the helper into the `ofxGui` namespace so that it doesn't appear in the general namespace, and to indicate that it belongs to `ofxGui`.

I've also added a dirty flag and checks for it so that the helper becomes easier to use - with this, calling `clear()` is not mandatory anymore which i hope makes its use a bit more intuitive.

arturoc commented:
Great thanks, can you remove the namespace though? we don't have namespaces anywhere else in the official addons so it would be weird to use it for this one class.

Also, and it probably doesn't make a big difference so feel free to leave it as it is. But it seems that most (all?) of the flags could just be replaced with a check to see if the mesh is empty which would make it even easier to follow the state? So just clear the mesh when something changes and then on draw if the mesh is empty regenerate it before draw which would avoid the boolean flags.

As i said apart from the namespace the other changes don't make such a big difference so feel free to leave it like it is if you prefer

arturoc commented:
Sorry I meant ofxGui::RectShape -> ofxGuiRectShape otherwise it would easily lead to clashes with some other library

arturoc commented:
thanks!
